### DISCUSSION TITLE
Naming convention (from the docs) -> name collision for composite unique constraints starting with same element

### DISCUSSION THREAD

**Discussion by lafrech** [sqlalchemy/sqlalchemy#10464] [Usage Questions] [answered]
I'm using the naming convention provided in the docs

- https://docs.sqlalchemy.org/en/20/core/constraints.html#configuring-constraint-naming-conventions
- https://alembic.sqlalchemy.org/en/latest/naming.html

```py
convention = {
    "ix": "ix_%(column_0_label)s",
    "uq": "uq_%(table_name)s_%(column_0_name)s",
    "ck": "ck_%(table_name)s_%(constraint_name)s",
    "fk": "fk_%(table_name)s_%(column_0_name)s_%(referred_table_name)s",
    "pk": "pk_%(table_name)s",
}
```

I just realized that the name of composite unique constraints only contains the name of the first element, which obviously means that two composite unique constraints starting with the same element end up having the same name.

```py
class Item(Base):
    __tablename__ = "items"
    __table_args__ = (
        sqla.UniqueConstraint("element_1", "element_2"),
        sqla.UniqueConstraint("element_1", "element_3"),
    )
```

When applying this to PostgreSQL, I get an error

```py
sqlalchemy.exc.ProgrammingError: (psycopg.errors.DuplicateTable)
```

I understand that the naming convention is a convenience and I can set names manually, but this came a  a surprise and I'm curious to know how people deal with this.

Also, would it be possible to define a name iterating other columns? (This might lead to long names, which raises other issues, such as the 63 characters limitation, at least in PostgreSQL.)

**Comment by zzzeek** [comment, marked as answer]
well the naming convention includes table name, column zero name, and that's it, so....does not seem surprising to me.     as you have surmised, for an uncommon case like this you can give explicit names to one or both of the unique constraints.

> Also, would it be possible to define a name iterating other composite keys? 

technically you can name "column_1" or any "column_X" in a naming constraint also, and I believe it returns a blank string for name not present, so you can do it that way


```py
from __future__ import annotations

from sqlalchemy import create_engine
from sqlalchemy import MetaData
from sqlalchemy import UniqueConstraint
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column


class Base(DeclarativeBase):
    metadata = MetaData(
        naming_convention={
            "ix": "ix_%(column_0_label)s",
            "uq": "uq_%(table_name)s_%(column_0_name)s_%(column_1_name)s",
            "ck": "ck_%(table_name)s_%(constraint_name)s",
            "fk": "fk_%(table_name)s_%(column_0_name)s_%(referred_table_name)s",
            "pk": "pk_%(table_name)s",
        }
    )


class Item(Base):
    __tablename__ = "items"
    id: Mapped[int] = mapped_column(primary_key=True)
    element_1: Mapped[int]
    element_2: Mapped[int]
    element_3: Mapped[int]

    __table_args__ = (
        UniqueConstraint("element_2"),
        UniqueConstraint("element_1", "element_2"),
        UniqueConstraint("element_1", "element_3"),
    )


e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

```

output

```
CREATE TABLE items (
	id INTEGER NOT NULL, 
	element_1 INTEGER NOT NULL, 
	element_2 INTEGER NOT NULL, 
	element_3 INTEGER NOT NULL, 
	CONSTRAINT pk_items PRIMARY KEY (id), 
	CONSTRAINT uq_items_element_2_ UNIQUE (element_2), 
	CONSTRAINT uq_items_element_1_element_2 UNIQUE (element_1, element_2), 
	CONSTRAINT uq_items_element_1_element_3 UNIQUE (element_1, element_3)
)

```

long names are truncated, but personally I'd use a hardcoded name for constraints like this

**Reply by lafrech** [reply]
> does not seem surprising to me

It's surprising for people like me who copied the convention without really trying to understand it, thinking "it's provided in the docs, let's not try to be clever". After all, I'm using SQLAlchemy and I've never had to look into the code, it just works (congratulations for that). Besides, it works well until the problem actually happens.

> technically you can name "column_1" or any "column_X" in a naming constraint also, and I believe it returns a blank string for name not present, so you can do it that way

Sure, if you do that to the 3rd or 4th element, you should be safe for a while, although the possibility of a 5-column-composite constraint exists. And you get trailing underscores which is just a cosmetic issue if an issue at all.

> long names are truncated

They are, and deterministically, but then unicity may be lost. (I faced that already, so now I try to stick to shorter names.)

Looks like you can't really set a naming convention and forget about it, which may be the reason this one is only suggested in the docs and not the default one.

> personally I'd use a hardcoded name for constraints like this

Fair enough. This is the kind of feedback I also wanted from this discussion.

Thanks.
